File: wallet/src/main/java/com/worldventures/wallet/analytics/general/SmartCardAnalyticErrorHandler.java
package com.worldventures.wallet.analytics.general;

import com.worldventures.wallet.analytics.WalletAnalyticsCommand;
import com.worldventures.wallet.analytics.general.action.SmartCardCommunicationErrorAction;
import com.worldventures.wallet.service.SmartCardErrorServiceWrapper;
import com.worldventures.wallet.service.WalletAnalyticsInteractor;
import com.worldventures.wallet.service.WalletAnalyticsServiceWrapper;

public class SmartCardAnalyticErrorHandler {

   private final WalletAnalyticsInteractor analyticsInteractor;

   private String currentNavigationState;

   public SmartCardAnalyticErrorHandler(SmartCardErrorServiceWrapper errorServiceWrapper,
         WalletAnalyticsServiceWrapper analyticsServiceWrapper, WalletAnalyticsInteractor analyticsInteractor) {
      this.analyticsInteractor = analyticsInteractor;

      errorServiceWrapper.addRequestFailureListener(this::trackError);
      analyticsServiceWrapper.addNavigationStateListener(state -> currentNavigationState = state);
   }

   private void trackError(Throwable t, String message) {
      if (currentNavigationState != null) {
         analyticsInteractor.walletAnalyticsPipe().send(new WalletAnalyticsCommand(
               new SmartCardCommunicationErrorAction(currentNavigationState, message)));
      }
   }
}

File: wallet/src/main/java/com/worldventures/wallet/analytics/general/SmartCardAnalyticErrorHandler.kt
package com.worldventures.wallet.analytics.general

import com.worldventures.wallet.analytics.WalletAnalyticsCommand
import com.worldventures.wallet.analytics.general.action.SmartCardCommunicationErrorAction
import com.worldventures.wallet.service.SmartCardErrorServiceWrapper
import com.worldventures.wallet.service.WalletAnalyticsInteractor
import com.worldventures.wallet.service.WalletAnalyticsServiceWrapper

class SmartCardAnalyticErrorHandler(errorServiceWrapper: SmartCardErrorServiceWrapper,
                                    analyticsServiceWrapper: WalletAnalyticsServiceWrapper, private val analyticsInteractor: WalletAnalyticsInteractor) {

   private var currentNavigationState: String? = null

   init {
      errorServiceWrapper.addRequestFailureListener { _, message -> this.trackError(message) }
      analyticsServiceWrapper.addNavigationStateListener { state -> currentNavigationState = state }
   }

   private fun trackError(message: String) {
      val state = currentNavigationState
      if (state != null) {
         analyticsInteractor.walletAnalyticsPipe().send(WalletAnalyticsCommand(
               SmartCardCommunicationErrorAction(state, message)))
      }
   }
}
